Source/MediaInfo/Audio/File_Usac.cpp

```diff
@@ -1803,16 +2252,485 @@ void File_Usac::UsacFrame(size_t BitsNotIncluded)
+
+    int16u N=0, lg=0;
+    int8u sampling_frequency_index_swb;
+    if (coreSbrFrameLengthIndex_Mapping[C.coreSbrFrameLengthIndex].coreCoderFrameLength==768 || coreSbrFrameLengthIndex_Mapping[C.coreSbrFrameLengthIndex].sbrRatioIndex)
+    {
+        float64 sampling_frequency_swb=Aac_sampling_frequency[C.sampling_frequency_index];
+        if (coreSbrFrameLengthIndex_Mapping[C.coreSbrFrameLengthIndex].coreCoderFrameLength==768)
+        {
+            //From specs: "For a transform length of 768 samples, the same 1024-based scalefactor band tables are used, but those corresponding to 4 / 3 * sampling frequency."
+            sampling_frequency_swb*=4;
+            sampling_frequency_swb/=3;
+        }
+        if (coreSbrFrameLengthIndex_Mapping[C.coreSbrFrameLengthIndex].sbrRatioIndex)
+        {
+            sampling_frequency_swb*=coreSbrFrameLengthIndex_Mapping[C.coreSbrFrameLengthIndex].coreCoderFrameLength/256;
+            sampling_frequency_swb/=coreSbrFrameLengthIndex_Mapping[C.coreSbrFrameLengthIndex].outputFrameLengthDivided256;
+        }
+        sampling_frequency_index_swb=Aac_AudioSpecificConfig_sampling_frequency_index(float64_int64s(sampling_frequency_swb));
+    }
+    else
+        sampling_frequency_index_swb=C.sampling_frequency_index;
+
+    if (num_windows==1)
+    {
+        if (C.coreSbrFrameLengthIndex<coreSbrFrameLengthIndex_Mapping_Size)
+            N=2*coreSbrFrameLengthIndex_Mapping[C.coreSbrFrameLengthIndex].coreCoderFrameLength;
+
+        if (max_sfb<=Aac_swb_offset_long_window[sampling_frequency_index_swb]->num_swb)
+            lg=Aac_swb_offset_long_window[sampling_frequency_index_swb]->swb_offset[max_sfb];
+    }
+    else
+    {
+        if (C.coreSbrFrameLengthIndex<coreSbrFrameLengthIndex_Mapping_Size)
+            N=2*coreSbrFrameLengthIndex_Mapping[C.coreSbrFrameLengthIndex].coreCoderFrameLength/8;
+
+        if (max_sfb<Aac_swb_offset_short_window[sampling_frequency_index_swb]->num_swb)
+            lg=Aac_swb_offset_short_window[sampling_frequency_index_swb]->swb_offset[max_sfb];
+    }
+
+    for (int8u win=0; win<num_windows; win++)
+        arithData(ch, N, lg, arith_reset_flag&&(win==0));
+
+    Element_End0();
+}
+
+
+//---------------------------------------------------------------------------
+void File_Usac::fdChannelStream(size_t ch, bool commonWindow, bool commonTw, bool tnsDataPresent, bool usacIndependencyFlag)
+{
+    Element_Begin1("fd_channel_stream");
+
+    Skip_S1(8,                                                  "global_gain");
+    if (C.noiseFilling)
+    {
+        Skip_S1(3,                                              "noise_level");
+        Skip_S1(5,                                              "noise_offset");
+    }
+
+    if (!commonWindow)
+        icsInfo();
+
+    if (C.tw_mdct && !commonTw)
+        twData();
+
+    scaleFactorData();
+
+    if (tnsDataPresent)
+    {
+        //TODO: tnsData();
+    }
+
+    acSpectralData(ch, usacIndependencyFlag);

```

Comment:
Add:

```

    if (!Aac_sampling_frequency[C.sampling_frequency_index])
    {
        Element_End0();
        return;
    }
```

(for smooth stop of parsing in case of unknown frequency index)